run-parts.8
.\" Hey, Emacs!  This is an -*- nroff -*- source file.
.\" Build-from-directory and this manpage are Copyright 1994 by Ian Jackson.
.\" Changes to this manpage are Copyright 1996 by Jeff Noxon.
.\" More
.\"
.\" This is free software; see the GNU General Public Licence version 2
.\" or later for copying conditions.  There is NO warranty.
.TH RUN\-PARTS 8 "14 Nov 2010" "Debian"
.SH NAME
run\-parts \- run scripts or programs in a directory
.SH SYNOPSIS
.PP
.B run\-parts
[\-\-test] [\-\-verbose] [\-\-report] [\-\-lsbsysinit] [\-\-regex=RE]
[\-\-umask=umask] [\-\-arg=argument] [\-\-exit\-on\-error] [\-\-help]
[\-\-version] [\-\-list] [\-\-reverse] [\-\-] DIRECTORY
.PP
.B run\-parts
\-V
.SH DESCRIPTION
.PP
.B run\-parts
runs all the executable files named within constraints described below, found
in directory
.IR directory .
Other files and directories are silently ignored.

If neither the \-\-lsbsysinit option nor the \-\-regex option is given
then the names must consist entirely of ASCII upper- and lower-case
letters, ASCII digits, ASCII underscores, and ASCII minus-hyphens.

If the \-\-lsbsysinit option is given, then the names must not end
in .dpkg\-old  or .dpkg\-dist or .dpkg\-new or .dpkg\-tmp, and must
belong to one or more of the following namespaces: the
LANANA-assigned namespace (^[a\-z0\-9]+$); the LSB hierarchical and
reserved namespaces (^_?([a\-z0\-9_.]+\-)+[a\-z0\-9]+$);
and the Debian cron script namespace (^[a\-zA-Z0\-9_\-]+$).

If the \-\-regex option is given, the names must match the custom
extended regular expression specified as that option's argument.

Files are run in the lexical sort order of their names unless the
\-\-reverse option is given, in which case they are run in the
opposite order.

.SH OPTIONS
.TP
.B \-\-test
print the names of the scripts which would be run, but don't actually run
them.
.TP
.B \-\-list
print the names of the all matching files (not limited to executables),
but don't actually run them. This option cannot be used with --test.
.TP
.B \-v, \-\-verbose
print the name of each script to stderr before running.
.TP
.B \-\-report
similar to
.BR \-\-verbose ,
but only prints the name of scripts which produce output.  The script's name is printed to whichever of stdout or stderr the script first produces output on.
.TP
.B \-\-reverse
reverse the scripts' execution order.
.TP
.B \-\-exit\-on\-error
exit as soon as a script returns with a non-zero exit code.
.TP
.B \-\-lsbsysinit
use LSB namespaces instead of classical behavior.
.TP
.B \-\-new\-session
run each script in a separate process session.  If you use this option,
killing run-parts will not kill the currently running script, it will
run until completion.
.TP
.BI \-\-regex= RE
validate filenames against custom extended regular expression
.IR RE .
See the EXAMPLES section for an example.
.TP
.BI "\-u, \-\-umask=" umask
sets the umask to
.I umask
before running the scripts.
.I umask
should be specified in octal.  By default the umask is set to 022.
.TP
.BI "\-a, \-\-arg=" argument
pass
.I argument
to the scripts.  Use
.B --arg
once for each argument you want passed.
.TP
.B "\-\-"
specifies that this is the end of the options.  Any filename after
.B "\-\-"
will be not be interpreted as an option even if it starts with a
hyphen.
.TP
.B \-h, \-\-help
display usage information and exit.
.TP
.B \-V, \-\-version
display version and copyright and exit.

.SH EXAMPLES
.P
Print the names of all files in /etc that start with `p' and end with `d':
.P
run-parts \-\-list \-\-regex \[aq]^p.*d$\[aq] /etc

.SH COPYRIGHT
.P
Copyright (C) 1994 Ian Jackson.
.P
Copyright (C) 1996 Jeff Noxon.
.P
Copyright (C) 1996, 1997, 1998 Guy Maor
.P
Copyright (C) 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009 Clint Adams

.B run\-parts
is free software; see the GNU General Public License version 2 or
later for copying conditions.  There is
.I no
warranty.


debian/changelog
debianutils (3.4.2) unstable; urgency=low

  * Add Slovene man page translations from Peter Kragelj.  closes:
    #598531.
  * French man page translation update from Christian Perrier.  closes:
    #597486.
  * run-parts.8: be more precise about ASCII requirement in naming.
    closes: #542029.

 -- Clint Adams <clint@gnu.org>  Sun, 14 Nov 2010 10:47:12 -0500

debianutils (3.4.1) unstable; urgency=low

  * po4a/po/fr.po: French man page translation update from
    Christian Perrier.  closes: #597430.
  * Bump to Standards-Version 3.9.1.

 -- Clint Adams <schizo@debian.org>  Sun, 19 Sep 2010 12:58:38 -0400

debianutils (3.4) unstable; urgency=low

  * Apply patch from Ariel to add strength options for compression and
    reduce xz default compression level.  closes: #587766.
  * Bump Standards-Version to 3.9.0.

 -- Clint Adams <schizo@debian.org>  Thu, 15 Jul 2010 22:02:27 -0400

debianutils (3.3) unstable; urgency=low

  * savelog.8: add SEE ALSO for logrotate(8).  closes: #487171.
  * savelog: add -J option for xz compression.  closes: #587520.

 -- Clint Adams <schizo@debian.org>  Wed, 30 Jun 2010 10:38:10 -0400

debianutils (3.2.3) unstable; urgency=low

  * Switch to 3.0 (native) source format.
  * French man page translation updates from Alexandre Hoïde.
    closes: #579652.
  * Bump to Standards-Version 3.8.4.

 -- Clint Adams <schizo@debian.org>  Thu, 29 Apr 2010 11:12:26 -0400

debianutils (3.2.2) unstable; urgency=low

  * savelog: patch from Patrick Coleman to fix breakage when target file
    does not exist.  closes: #556227.
  * Bump to Standards-Version 3.8.3.

 -- Clint Adams <schizo@debian.org>  Sat, 14 Nov 2009 13:16:01 -0500

debianutils (3.2.1) unstable; urgency=low

  * run-parts.8: conform man page regex with the actual regex used by
    run-parts.  closes: #536395.

 -- Clint Adams <schizo@debian.org>  Sun, 09 Aug 2009 15:18:52 -0400

debianutils (3.2) unstable; urgency=low

  * run-parts.c: print command arguments when --verbose and --arg are
    used.  closes: #535621.
  * Bump to Standards-Version 3.8.2.

 -- Clint Adams <schizo@debian.org>  Sun, 05 Jul 2009 15:03:00 -0400

debianutils (3.1.3) unstable; urgency=low

  * Remove any mention of readlink from debian/copyright.  closes:
    #527138.

 -- Clint Adams <schizo@debian.org>  Tue, 05 May 2009 21:08:31 -0400

debianutils (3.1.2) unstable; urgency=low

  * Patch from Zack Weinberg to fix tempfile -n crash.  closes: #526981.

 -- Clint Adams <schizo@debian.org>  Mon, 04 May 2009 18:56:14 -0400

debianutils (3.1.1) unstable; urgency=medium

  [ Manoj Srivastava ]
  * Fix a double free bug in tempfile. Closes: #526606, #526607, #526610

 -- Manoj Srivastava <srivasta@debian.org>  Sat, 02 May 2009 03:45:30 -0500

debianutils (3.1) unstable; urgency=medium

  [ Manoj Srivastava ]
  * which.1: clarify that which does not follow symlinks.  closes: #268624.
  * run-parts.c: permission and sanity checks run on link targets, not
    symlinks.  closes: #294789.
  * tempfile.c: tempfile with a suffix no longer needs hardlinks.
    closes: #312270.
  * savelog.8: disambiguate version numbering in manual.  closes: #330289.
  * tempfile.1: update manual page with up-to-date tempnam behaviors.
    closes: #338490.

  [ Clint Adams ]
  * savelog.8, tempfile.1: remove spurious trailing whitespace
  * savelog.8, tempfile.1, which.1: update dates in manual pages.
  * po4a/po/fr.po: tweak French translations enough to reach 80%.

 -- Clint Adams <schizo@debian.org>  Fri, 01 May 2009 18:01:28 -0400

debianutils (3.0.2) unstable; urgency=low

  * installkernel: fix symlinking of kernel images.  closes: #526493.

 -- Clint Adams <schizo@debian.org>  Fri, 01 May 2009 11:08:13 -0400

debianutils (3.0.1) unstable; urgency=low

  * Remove French sensible-editor man page as well.  closes: #521543.

 -- Clint Adams <schizo@debian.org>  Sun, 29 Mar 2009 17:55:21 -0400

debianutils (3.0) unstable; urgency=low

  * Stop shipping sensible-browser, sensible-editor, sensible-pager, and
    accompanying man pages, and declare a dependency on sensible-utils,
    to which they have moved.

 -- Clint Adams <schizo@debian.org>  Thu, 26 Mar 2009 21:25:08 -0400

debianutils (2.32.3) unstable; urgency=medium

  * installkernel: remove stray right bracket, thanks to Alban Browaeys.
    closes: #521286.

 -- Clint Adams <schizo@debian.org>  Thu, 26 Mar 2009 09:53:46 -0400

debianutils (2.32.2) unstable; urgency=low

  * tempfile.c: patch from Ryan Niebur to fix breakage of -n option.
    closes: #520940.
  * tempfile.c: minor tweaks.

 -- Clint Adams <schizo@debian.org>  Mon, 23 Mar 2009 21:45:08 -0400

debianutils (2.32.1) unstable; urgency=low

  [ Manoj Srivastava ]
  * Ignore a fifth argument to installkernel.  closes: #494861.

  [ Clint Adams ]
  * installkernel: patch from Thibaut VARENE to fix symlinks better.
    closes: #520690.

 -- Clint Adams <schizo@debian.org>  Sun, 22 Mar 2009 18:58:46 -0400

debianutils (2.32) unstable; urgency=low

  [ Clint Adams ]
  * tempfile.c: free memory right before exit, thanks to Franck
    Joncourt.
  * Bump to Standards-Version 3.8.1.

  [ Manoj Srivastava ]
  * [654eca6]: Remove mkboot entirely
    mkboot has become increasingly distant from accepted practices on
    modern Debian isntallations, and has limited utility. mkboot only
    works on i386 and ia64. If GRUB is installed, it does nothing. It is
    unsure how the grub2 detection works. mkboot only works on floppy
    diskette drives, which are increasingly rare beasts. If rdev is not
    present, mkboot will fail to determine the correct root partition. In
    most cases, running mkboot is the wrong thing to do, and this
    bugginess, in conjunction with its irrelevancy, require that this be
    removed.
    Bug fix: "mkboot cannot make boot floppy without lilo installed, but
    package has no dependency on lilo.", thanks to Chris Metzler
    (Closes:  #197463).
    Bug fix: "try different floppy formats when linux is larger than

run-parts.c
/* run-parts: run a bunch of scripts in a directory
 *
 * Debian run-parts program
 * Copyright (C) 1996 Jeff Noxon <jeff@router.patch.net>,
 * Copyright (C) 1996-1999 Guy Maor <maor@debian.org>
 * Copyright (C) 2002, 2003, 2004, 2005 Clint Adams <schizo@debian.org>
 *
 * This is free software; see the GNU General Public License version 2
 * or later for copying conditions.  There is NO warranty.
 *
 * Based on run-parts.pl version 0.2, Copyright (C) 1994 Ian Jackson.
 *
 */

#include <stdio.h>
#include <stdarg.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <dirent.h>
#include <sys/stat.h>
#include <unistd.h>
#include <fcntl.h>
#ifdef HAVE_GETOPT_H
#include <getopt.h>
#endif /* HAVE_GETOPT_H */
#include <string.h>
#include <errno.h>
#include <ctype.h>
#include <signal.h>
#include <sys/time.h>
#include <regex.h>

#define RUNPARTS_NORMAL 0
#define RUNPARTS_ERE 1
#define RUNPARTS_LSBSYSINIT 100

int test_mode = 0;
int list_mode = 0;
int verbose_mode = 0;
int report_mode = 0;
int reverse_mode = 0;
int exitstatus = 0;
int regex_mode = 0;
int exit_on_error_mode = 0;
int new_session_mode = 0;

int argcount = 0, argsize = 0;
char **args = 0;

char *custom_ere;
regex_t hierre, tradre, excsre, classicalre, customre;

static char* regex_get_error(int errcode, regex_t *compiled);
static void  regex_compile_pattern(void);
static void  regex_clean(void);

void error(char *format, ...)
{
  va_list ap;

  fprintf(stderr, "run-parts: ");

  va_start(ap, format);
  vfprintf(stderr, format, ap);
  va_end(ap);

  fprintf(stderr, "\n");
}


void version()
{
  fprintf(stderr, "Debian run-parts program, version " PACKAGE_VERSION
	  "\nCopyright (C) 1994 Ian Jackson, Copyright (C) 1996 Jeff Noxon.\n"
	  "Copyright (C) 1996,1997,1998,1999 Guy Maor\n"
	  "Copyright (C) 2002, 2003, 2004, 2005, 2006, 2007 Clint Adams\n"
	  "This is free software; see the GNU General Public License version 2\n"
	  "or later for copying conditions.  There is NO warranty.\n");
  exit(0);
}


void usage()
{
  fprintf(stderr, "Usage: run-parts [OPTION]... DIRECTORY\n"
	  "      --test          print script names which would run, but don't run them.\n"
	  "      --list          print names of all valid files (can not be used with\n"
	  "                      --test)\n"
	  "  -v, --verbose       print script names before running them.\n"
	  "      --report        print script names if they produce output.\n"
	  "      --reverse       reverse execution order of scripts.\n"
	  "      --exit-on-error exit as soon as a script returns with a non-zero exit\n"
	  "                      code.\n"
	  "      --lsbsysinit    validate filenames based on LSB sysinit specs.\n"
	  "      --new-session   run each script in a separate process session\n"
	  "      --regex=PATTERN validate filenames based on POSIX ERE pattern PATTERN.\n"
	  "  -u, --umask=UMASK   sets umask to UMASK (octal), default is 022.\n"
	  "  -a, --arg=ARGUMENT  pass ARGUMENT to scripts, use once for each argument.\n"
	  "  -V, --version       output version information and exit.\n"
	  "  -h, --help          display this help and exit.\n");
  exit(0);
}


/* The octal conversion in libc is not foolproof; it will take the 8 and 9
 * digits under some circumstances.  We'll just have to live with it.
 */
void set_umask()
{
  int mask, result;

  result = sscanf(optarg, "%o", &mask);
  if ((result != 1) || (mask > 07777) || (mask < 0)) {
    error("bad umask value");
    exit(1);
  }

  umask(mask);
}

/* Add an argument to the commands that we will call.  Called once for
   every argument. */
void add_argument(char *newarg)
{
  if (argcount + 1 >= argsize) {
    argsize = argsize ? argsize * 2 : 4;
    args = realloc(args, argsize * (sizeof(char *)));
    if (!args) {
      error("failed to reallocate memory for arguments: %s", strerror(errno));
      exit(1);
    }
  }
  args[argcount++] = newarg;
  args[argcount] = 0;
}

/* True or false? Is this a valid filename? */
int valid_name(const struct dirent *d)
{
    char         *s;
    unsigned int  retval;

    s = (char *)&(d->d_name);

    if (regex_mode == RUNPARTS_ERE)
        retval = !regexec(&customre, s, 0, NULL, 0);

    else if (regex_mode == RUNPARTS_LSBSYSINIT) {

        if (!regexec(&hierre, s, 0, NULL, 0))
            retval = regexec(&excsre, s, 0, NULL, 0);

	else
            retval = !regexec(&tradre, s, 0, NULL, 0);

    } else
        retval = !regexec(&classicalre, s, 0, NULL, 0);

    return retval;
}

/* Execute a file */
void run_part(char *progname)
{
  int result;
  int pid;
  int pout[2], perr[2];

  if (report_mode && (pipe(pout) || pipe(perr))) {
    error("pipe: %s", strerror(errno));
    exit(1);
  }
  if ((pid = fork()) < 0) {
    error("failed to fork: %s", strerror(errno));
    exit(1);
  }
  else if (!pid) {
    if (new_session_mode)
      setsid();
    if (report_mode) {
      if (dup2(pout[1], STDOUT_FILENO) == -1 ||
	  dup2(perr[1], STDERR_FILENO) == -1) {
	error("dup2: %s", strerror(errno));
	exit(1);
      }
      close(pout[0]);
      close(perr[0]);
      close(pout[1]);
      close(perr[1]);
    }
    args[0] = progname;
    execv(progname, args);
    error("failed to exec %s: %s", progname, strerror(errno));
    exit(1);
  }

  if (report_mode) {
    fd_set set;
    int max, r, printflag;
    ssize_t c;
    char buf[4096];

    close(pout[1]);
    close(perr[1]);
    max = pout[0] > perr[0] ? pout[0] + 1 : perr[0] + 1;
    printflag = 0;

    while (pout[0] >= 0 || perr[0] >= 0) {

      do {
	FD_ZERO(&set);
	if (pout[0] >= 0)
	  FD_SET(pout[0], &set);
	if (perr[0] >= 0)
	  FD_SET(perr[0], &set);
	r = select(max, &set, 0, 0, 0);
      } while (r < 0 && errno == EINTR);

      if (r < 0) {
	/* assert(errno != EINTR) */
	error("select: %s", strerror(errno));
	exit(1);
      }
      else if (r > 0) {
	if (pout[0] >= 0 && FD_ISSET(pout[0], &set)) {
	  c = read(pout[0], buf, sizeof(buf));
	  if (c > 0) {
	    if (!printflag) {
	      printf("%s:\n", progname);
	      fflush(stdout);
	      printflag = 1;
	    }
	    write(STDOUT_FILENO, buf, c);
	  }
	  else if (c == 0) {
	    close(pout[0]);
	    pout[0] = -1;
	  }
	  else if (c < 0) {
	    close(pout[0]);
	    pout[0] = -1;
	    error("failed to read from stdout pipe: %s", strerror (errno)); 
	  }
	}
	if (perr[0] >= 0 && FD_ISSET(perr[0], &set)) {
	  c = read(perr[0], buf, sizeof(buf));
	  if (c > 0) {
	    if (!printflag) {
	      fprintf(stderr, "%s:\n", progname);
	      fflush(stderr);
	      printflag = 1;
	    }
	    write(STDERR_FILENO, buf, c);
	  }
	  else if (c == 0) {
	    close(perr[0]);
	    perr[0] = -1;
	  }
	  else if (c < 0) {
	    close(perr[0]);
	    perr[0] = -1;
	    error("failed to read from error pipe: %s", strerror (errno)); 
	  }
	}
      }
      else {
	/* assert(FALSE): select was called with infinite timeout, so
	   it either returns successfully or is interrupted */
      }				/*if */
    }				/*while */
  }

  waitpid(pid, &result, 0);

  if (WIFEXITED(result) && WEXITSTATUS(result)) {
    error("%s exited with return code %d", progname, WEXITSTATUS(result));
    exitstatus = 1;
  }
  else if (WIFSIGNALED(result)) {
    error("%s exited because of uncaught signal %d", progname,
	  WTERMSIG(result));
    exitstatus = 1;
  }
}


/* Find the parts to run & call run_part() */
void run_parts(char *dirname)
{
  struct dirent **namelist;
  char *filename;
  size_t filename_length, dirname_length;
  int entries, i, result;
  struct stat st;

  /* dirname + "/" */
  dirname_length = strlen(dirname) + 1;
  /* dirname + "/" + ".." + "\0" (This will save one realloc.) */
  filename_length = dirname_length + 2 + 1;
  if (!(filename = malloc(filename_length))) {
    error("failed to allocate memory for path: %s", strerror(errno));
    exit(1);
  }
  strcpy(filename, dirname);
  strcat(filename, "/");

  /* scandir() isn't POSIX, but it makes things easy. */
  entries = scandir(dirname, &namelist, valid_name, alphasort);
  if (entries < 0) {
    error("failed to open directory %s: %s", dirname, strerror(errno));
    exit(1);
  }

  i = reverse_mode ? 0 : entries;
  for (i = reverse_mode ? (entries - 1) : 0;
       reverse_mode ? (i >= 0) : (i < entries); reverse_mode ? i-- : i++) {
    if (filename_length < dirname_length + strlen(namelist[i]->d_name) + 1) {
      filename_length = dirname_length + strlen(namelist[i]->d_name) + 1;
      if (!(filename = realloc(filename, filename_length))) {
	error("failed to reallocate memory for path: %s", strerror(errno));
	exit(1);
      }
    }
    strcpy(filename + dirname_length, namelist[i]->d_name);

    strcpy(filename, dirname);
    strcat(filename, "/");
    strcat(filename, namelist[i]->d_name);

    result = stat(filename, &st);
    if (result < 0) {
      error("failed to stat component %s: %s", filename, strerror(errno));
      if (exit_on_error_mode) {
        exit(1);
      }
      else
        continue;
    }

    if (S_ISREG(st.st_mode)) {
      if (!access(filename, X_OK)) {
	if (test_mode) {
	  printf("%s\n", filename);
	}
	else if (list_mode) {
	  if (!access(filename, R_OK))
	    printf("%s\n", filename);
	}
	else {
	  if (verbose_mode)
	    if (argcount) {
	      char **a = args;

	      fprintf(stderr, "run-parts: executing %s", filename);
	      while(*(++a))
		fprintf(stderr, " %s", *a);
	      fprintf(stderr, "\n");
	    } else {
	      fprintf(stderr, "run-parts: executing %s\n", filename);
	    }
	  run_part(filename);
	  if (exitstatus != 0 && exit_on_error_mode) return;
	}
      }
      else if (!access(filename, R_OK)) {
	if (list_mode) {
	  printf("%s\n", filename);
	}
      }
      else if (S_ISLNK(st.st_mode)) {
	if (!list_mode) {
	  error("run-parts: component %s is a broken symbolic link\n",filename);
	  exitstatus = 1;
	}
      }
    }
    else if (!S_ISDIR(st.st_mode)) {
      if (!list_mode) {
	error("run-parts: component %s is not an executable plain file\n",
	       filename);
	exitstatus = 1;
      }
    }

    free(namelist[i]);
  }
  free(namelist);
  free(filename);
}

/* Process options */
int main(int argc, char *argv[])
{
  custom_ere = NULL;
  umask(022);
  add_argument(0);

  for (;;) {
    int c;
    int option_index = 0;

    static struct option long_options[] = {
      {"test", 0, &test_mode, 1},
      {"list", 0, &list_mode, 1},
      {"verbose", 0, 0, 'v'},
      {"report", 0, &report_mode, 1},
      {"reverse", 0, &reverse_mode, 1},
      {"umask", 1, 0, 'u'},
      {"arg", 1, 0, 'a'},
      {"help", 0, 0, 'h'},
      {"version", 0, 0, 'V'},
      {"lsbsysinit", 0, &regex_mode, RUNPARTS_LSBSYSINIT},
      {"regex", 1, &regex_mode, RUNPARTS_ERE},
      {"exit-on-error", 0, &exit_on_error_mode, 1},
      {"new-session", 0, &new_session_mode, 1},
      {0, 0, 0, 0}
    };

    c = getopt_long(argc, argv, "u:ha:vV", long_options, &option_index);
    if (c == EOF)
      break;
    switch (c) {
    case 0:
      if(option_index==10) { /* hardcoding this will lead to trouble */
        custom_ere = strdup(optarg);
      }
      break;
    case 'u':
      set_umask();
      break;
    case 'a':
      add_argument(optarg);
      break;
    case 'h':
      usage();
      break;
    case 'v':
      verbose_mode = 1;
      break;
    case 'V':
      version();
      break;
    default:
      fprintf(stderr, "Try `run-parts --help' for more information.\n");
      exit(1);
    }
  }

  /* We require exactly one argument: the directory name */
  if (optind != (argc - 1)) {
    error("missing operand");
    fprintf(stderr, "Try `run-parts --help' for more information.\n");
    exit(1);
  } else if (list_mode && test_mode) {
    error("--list and --test can not be used together");
    fprintf(stderr, "Try `run-parts --help' for more information.\n");
    exit(1);
  } else {
    regex_compile_pattern();
    run_parts(argv[optind]);
    regex_clean();

    free(args);
    free(custom_ere);

    return exitstatus;
  }
}

/*
 * Compile patterns used by the application
 *
 * In order for a string to be matched by a pattern, this pattern must be
 * compiled with the regcomp function. If an error occurs, the application
 * exits and displays the error.
 */
static void
regex_compile_pattern (void)
{
    int      err;
    regex_t *pt_regex;

    if (regex_mode == RUNPARTS_ERE) {

        if ((err = regcomp(&customre, custom_ere,
                    REG_EXTENDED | REG_NOSUB)) != 0)
            pt_regex = &customre;

    } else if (regex_mode == RUNPARTS_LSBSYSINIT) {

        if ( (err = regcomp(&hierre, "^_?([a-z0-9_.]+-)+[a-z0-9]+$",
                    REG_EXTENDED | REG_NOSUB)) != 0)
            pt_regex = &hierre;

        else if ( (err = regcomp(&excsre, "^[a-z0-9-].*dpkg-(old|dist|new|tmp)$",
                    REG_EXTENDED | REG_NOSUB)) != 0)
            pt_regex = &excsre;

        else if ( (err = regcomp(&tradre, "^[a-z0-9][a-z0-9-]*$", REG_NOSUB))
                    != 0)
            pt_regex = &tradre;

    } else if ( (err = regcomp(&classicalre, "^[a-zA-Z0-9_-]+$",
                    REG_EXTENDED | REG_NOSUB)) != 0)
        pt_regex = &classicalre;

    if (err != 0) {
        fprintf(stderr, "Unable to build regexp: %s", \
                            regex_get_error(err, pt_regex));
        exit(1);
    }
}

/*
 * Get a regex error.
 *
 * This function allocates a buffer to store the regex error description.
 * If a buffer cannot be allocated, then the use of xmalloc will end the
 * program.
 *
 * @errcode: return error code from a one of the regex functions
 * @compiled: compile pattern which causes the failure
 *
 * It returns a pointer on the current regex error description.
 */
static char *
regex_get_error(int errcode, regex_t *compiled)
{
    size_t  length;
    char     *buf;

    length = regerror(errcode, compiled, NULL, 0);
    buf    = malloc(length);
    if (buf == 0) {
        error("Virtual memory exhausted\n");
        exit(1);
    }

    regerror(errcode, compiled, buf, length);

    return buf;
}

/*
 * Clean the compiled patterns according to the current regex_mode
 */
static void
regex_clean(void)
{
    if (regex_mode == RUNPARTS_ERE)
        regfree(&customre);

    else if (regex_mode == RUNPARTS_LSBSYSINIT) {
        regfree(&hierre);
        regfree(&excsre);
        regfree(&tradre);

    } else
        regfree(&classicalre);
}
